### ISSUE TITLE
Abandoned project ?

### ISSUE THREAD

**Issue by FrankHovis** [LaserWeb/LaserWeb4-Binaries#37]
Its over 2 years now since any activity here, has this been abandoned?

**Comment by dgoadby** [comment]
It seems so. I use this so did wonder if Frank would agree to it being forked or transferred to someone who is active in the group. There are a few bugs to clear up and some improvement test I would want.

**Comment by cprezzi** [comment]
The frontend devs have left the project, but I could give new devs acces to this repo, if someone is capable of taking over.
The frontend is based on react and redux.

**Comment by webii** [comment]
I am a fullstack dev and using this software a lot. As I am using raspberry pi zero w, LaserWeb is a perfect solution for remote working. But there are a few bugs which are really painful. I would like to fix them if I can get at least some support and code review.

**Comment by cprezzi** [comment]
I am open for pull requests and will support you as good as I can. Just tell me what you need to know.

**Comment by webii** [comment]
Is there any way to contact you more directly, like using private messages or discord?
First I started looking on server code. But before any serious changes I needed to do some clean up, to remove duplicated code :/ https://github.com/webii/lw.comm-server/pull/1

**Comment by cprezzi** [comment]
I am still maintaining the server code. What we realy need are frontend devs. 

I saw your refactoring, which chages a lot of code according to your personal preferences. Are you willing to take over backend maintenance for the the next few years? 

You also removed the App folder, which is the hosted frontend. This is needed for server only installations (like RPi) which are used by browser over network. Did you consider all usecases (server only, server & app, app only)?

**Comment by webii** [comment]
1. There is no change in logic in my MR.
2. Changes are not due to my preference, but due to the good practices, as code there were duplicated three times, and still in many places is. Because for each type of connection there was the same (almost) code. It is hardly manageable.
3. Fronted app is not needed there. It could be installed by downloading from the other repository. I will add a message in the readme for those who need hosting also frontend. As someone is working on raspberry pi this frontend is not needed as you can connect from any frontend (electron) providing raspberry pi address and port. Also providing frontend on the server side might also be misleading, because there is no way to use laser cutter for more than one person/session, also the web session cannot be shared.
4. I have removed linux binary. Those binaries could be added as artifacts when building specific release, they don't have to be part of a code repository.

**Comment by cprezzi** [comment]
1. I agree.
2. I partly agree but using let instead of var (ts over vanilla) is just personal preference.
3. Many users have a pi as the server and don't want to install an app for the frontend but instead just use a webbrowser on a PC or tablet. The hosted app is also needed in the case when the electron exe is run as server and the frontend wants to be used by webbrowser. By the way: It IS possible to connect multiple frontends to the same server. Each frontend will show the actual position and can pause, resume, stop the job or change feed and spindle. Even the whole gcode is sent to second frontend, it just cant show the job jet.
4. I agree.

**Comment by helifino** [comment]
However, now there's no linux build in Releases. For someone trying to install to Linux desktop, but not comfortable building from sources, this is unusable.

**Comment by cprezzi** [comment]
@helifino The installation process for Linux is described on https://github.com/LaserWeb/lw.comm-server/wiki/Manual-Installation-(RasPi)

**Comment by helifino** [comment]
Please then update the Installation (Linux) page. https://laserweb.yurl.ch/documentation/installation/34-install-linux-appimage-format

**Comment by cprezzi** [comment]
Thanks for that hint! I completely forgot this page.

**Comment by helifino** [comment]
> The installation process for Linux is described on https://github.com/LaserWeb/lw.comm-server/wiki/Manual-Installation-(RasPi)

This fails the no-coding-knowledge test. Had to determine and solve several dependency issues just to install, and still it does nothing. Connect via browser gives 'connecting to server' in the status pane but no connection. Linux Mint 20.3

**Comment by cprezzi** [comment]
@helifino Did you get the frontend in the browser?
Did you set the correct server connection? (Go to the comms tab, click "Server Connection" edit "Server IP" to own IP:8000 and hit connect)
What was shown in the console where you started the server.js?

**Comment by helifino** [comment]
Oddly, without changing anything, now it's actually connecting to itself but not to the laser:
![image](https://user-images.githubusercontent.com/8587528/164998139-3334fd5c-83db-4d87-a69f-f3555b4c9edd.png)

**Comment by cprezzi** [comment]
It seems like server.js does not have the right to open /dev/ttyUSB0. Are you sure that this is the correct interface? On Raspberry Pi, it is /dev/ttyAMA0. 
You could also try if it works when you start the server app with superuser rights (sudo).

**Comment by helifino** [comment]
Yes, this is not a pi.

**Comment by helifino** [comment]
Point of clarification: Desktop version can communicate with machine. Comms server cannot. Launched from same user account.

**Comment by cprezzi** [comment]
Unfortunately I am not a Linux guy, so I can't realy help. I just remember that another user pointed out that Linux is very restrictive with access to ressources and the right to acces ttyUSB0 needs to be explicitly configured. 

I know that you are not using a pi, I just mentioned that on a pi it would be ttyAMA0. I don't know what is correct on your Linux.

**Comment by neonflux-io** [comment]
Hi @cprezzi, I’d be interested in helping with the LaserWeb frontend and potentially taking on longer-term maintenance.

I have strong experience with React, Redux, JavaScript/TypeScript, UI/UX, and full-stack development. I’m happy to start by fixing existing frontend bugs, improving maintainability, and reviewing the current build/release workflow.

If you’re still looking for frontend developers, I’d be glad to contribute.

**Comment by webii** [comment]
@neonflux-io I am also open to cooperation, there is already my PR with some cleaning, with some help it can be tested, reviewed and maybe it could be a good starter to further works. It was quite long ago, nowadays with help of LLM it should be easier to refactor this. In general it would be great to have this kind of tool for laser cutting.

**Comment by cprezzi** [comment]
Hi @neonflux-io and @webii 

The core devs of this project moved away log time ago. I stay to coordinate new help, which is very welcome, to move LaserWeb forward. It's just important that all usecases are known and tested after bigger changes. The usecases are:
- Standalone App (Electron EXE) with embedded server and client
- Standalone App used as Server (for machine connection) and remote Client by Webbrowser or standalone App
- NodeJS-Server on Linux/RaspberryPi and remote Client by Webbrowser or standalone App.
- Installation as Docker or ...

There is already a partial overhaul in a fork from @easytarget and some work from @joepie91 that you should check.

If you are commited for longer term maintenance, I can give you access to the LaserWeb Repositories to directly push or merge your PRs yourself.

Just tell me, if you are interested.

**Comment by neonflux-io** [comment]
Hi, @cprezzi 

I am very interested.
Can we discuss in whatsapp?
This is my whatsapp number: +353 83 098 1869

I will wait for your message.

**Comment by webii** [comment]
To have this project working well is important to separate components, so separate backend and frontend. Also this is important to have some abstraction layer, to not fall into specific implementation.

Use case in terms of product development is laser cutting actually. Those exe, dmgs or dockers are nonfunctional requirements, they are only wrappers or runtimes to running this, which actually with help of the LLM is the easiest part.

For me, also this is important to add at least some testing to backend / frontend / e2w as manual testing with every change is just a nightmare.

**Comment by cprezzi** [comment]
@webii 
I don't agree with a complete separation of backend and frontend. The large majority of LaserWeb users use it as a standalone Application on Windows or Mac (frontend with integrated backend and local machine connection). Separate client & server is only used in very rare cases, linke Raspberry Pi as remote machine interface and client from another pc.

Technically the server part (backend lw.comm-server) is already separated from the frontend (LaserWeb4) and both can be run separately. There is only an additional step to generate the all-in-one App.

**Comment by cprezzi** [comment]
Here a link to an older discussion with a lot of infos about the separate frontend and backend and the process to build the combined electron app: https://github.com/LaserWeb/LaserWeb4/issues/562#issuecomment-538924859

The simplest way to start is to clone LaserWeb/lw.comm-server, then run the node server.js and call the frontend via Webbrowser (serverIP:8000). If that works, you can move to the LaserWeb4/dev-es6 Branch, start "npm start-app" (webpack server) while in paralell changing sourcecode. The Webpack Server will recompile on the fly after each change and reload the Website.

**Comment by cprezzi** [comment]
If that works, you can move to the LaserWeb4/dev-es6 Branch, start the webpack server with "npm start-app", which should "compile" the frontend and open a browser on it. Now you can start coding. The Webpack Server will recompile on the fly after each save and reload the Website.

**Comment by webii** [comment]
I used word component separation, they should be separate in the code with some interfaces between those two components. But also with a possibility to run only one of them. Also it is important to have some code architecture to not end up with spaghetti.

**Comment by cprezzi** [comment]
@webii I don't understand what you mean? Please read https://github.com/LaserWeb/lw.comm-server/wiki to better understand the conzept of LaserWeb backend and frontend.
